### DISCUSSION TITLE
Sending large HTTP responses without io_buf

### DISCUSSION THREAD

**Discussion by hoeken** [cesanta/mongoose#2481] [Ask questions, get answers] [answered]
I'm working on converting some code to mongoose, but I'm running into a major issue.  I can't seem to find a way to send large responses that bypass the IO buffers.  I am on esp32 and am having issues with heap size.

I have a large (133380 byte) index.html file that is processed and included as a header file like so:

`#define index_html_gz_len 133380
const uint8_t index_html_gz[] PROGMEM = {
0x1f,0x8b,0x08,0x00,0x00,0x00,0x00,0x00,0x00,0x13......}`

Unfortunately when I go to send this file with any of the mg_http_reply, mg_send, or mg_printf functions I run into issues with allocating more heap:

`462d   1 mongoose.c:2840:mg_iobuf_resiz 2048->135168`

Just before this call, I checked the heap with **DUMP(heap_caps_get_largest_free_block(MALLOC_CAP_DEFAULT));** and get something like this.

`MongooseHttpServer.cpp:677: heap_caps_get_largest_free_block((1<<12)) = 110580`

So clearly my issue is that it cannot allocate a bigger send buffer, but really this file already exists as a stream and doesn't need to be copied, just sent out.  Anyone have any suggestions?

**Comment by scaprile** [comment, marked as answer]
Please follow the guidelines in the [documentation](https://mongoose.ws/documentation),  [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) available.
https://mongoose.ws/documentation/tutorials/huge-response/
or
This is TCP, it is streaming, just split your whatever in pieces that will fit, or, put your file in a file system and let the file system do your job.
